modules/adagioBidAdapter.js

```diff
@@ -1,73 +1,355 @@
 import find from 'core-js/library/fn/array/find';
 import * as utils from '../src/utils';
-import { registerBidder } from '../src/adapters/bidderFactory';
+import {registerBidder} from '../src/adapters/bidderFactory';
 
 const BIDDER_CODE = 'adagio';
-const VERSION = '1.0.0';
+const VERSION = '2.0.0';
+const FEATURES_VERSION = '1';
 const ENDPOINT = 'https://mp.4dex.io/prebid';
 const SUPPORTED_MEDIA_TYPES = ['banner'];
+const ADAGIO_TAG_URL = '//script.4dex.io/localstore.js';
+const ADAGIO_TAG_TO_LOCALSTORE = '//script.4dex.io/adagio.js';
+const ADAGIO_LOCALSTORE_KEY = 'adagioScript';
+const LOCALSTORE_TIMEOUT = 100;
+const ADSRV_EVENTS = {
+  GPT: {
+    IMPRESSION_VIEWABLE: 'impressionViewable',
+    SLOT_ON_LOAD: 'slotOnLoad',
+    SLOT_RENDER_ENDED: 'slotRenderEnded',
+    SLOT_REQUESTED: 'slotRequested',
+    SLOT_RESPONSE_RECEIVED: 'slotResponseReceived',
+    SLOT_VISIBILITY_CHANGED: 'slotVisibilityChanged',
+  },
+  SAS: {
+    CALL: 'call',
+    CLEAN: 'clean',
+    BEFORE_RENDER: 'beforeRender',
+    CMP_ANSWERED: 'CmpAnswered',
+    CMP_CALLED: 'CmpCalled',
+    LOAD: 'load',
+    NOAD: 'noad',
+    RENDER: 'render',
+    RESET: 'reset'
+  }
+};
 
-/**
- * Based on https://github.com/ua-parser/uap-cpp/blob/master/UaParser.cpp#L331, with the following updates:
- * - replaced `mobile` by `mobi` in the table regexp, so Opera Mobile on phones is not detected as a tablet.
- */
-function _getDeviceType() {
-  let ua = navigator.userAgent;
+function canAccessTopWindow() {
+  try {
+    if (utils.getWindowTop().location.href) {
+      return true;
+    }
+  } catch (error) {
+    return false;
+  }
+}
+
+function initAdagio() {
+  const script = document.createElement('script');
+  const w = utils.getWindowTop();
+
+  w.ADAGIO = w.ADAGIO || {};
+  w.ADAGIO.queue = w.ADAGIO.queue || [];
+  w.ADAGIO.versions = w.ADAGIO.versions || {};
+  w.ADAGIO.versions.adagioBidderAdapter = VERSION;
+
+  const getAdagioTag = function getAdagioTag() {
+    const ls = w.localStorage.getItem('adagioScript');
+    if (ls !== null) {
+      Function(ls)(); // eslint-disable-line no-new-func
+    } else {
+      utils.logWarn('Adagio Script not found');
+    }
+  }
+
+  const adagioEnqueue = function adagioEnqueue(action, data) {
+    w.ADAGIO.queue.push({ action, data, ts: Date.now() });
+  }
+
+  // Listen to ad-server events in current window as we can
+  // in a Post-Bid scenario.
+  window.googletag = window.googletag || {};
+  window.googletag.cmd = window.googletag.cmd || [];
+  window.googletag.cmd.push(function() {
+    const gptEvents = Object.keys(ADSRV_EVENTS.GPT).map(key => ADSRV_EVENTS.GPT[key]);
+    gptEvents.forEach(eventName => {
+      window.googletag.pubads().addEventListener(eventName, args => {
+        adagioEnqueue('gpt-event', { eventName, args });
```

Comment:
what are you doing with the event data?

Reply:
We listen to the ad-server events to know when an ad is visually rendered on the page and to start a measurer (goal of our external script)
We also use the ad-server event information like isEmpty, returned size, etc to improve/compare our measurer data

We made the choice to listen to ad-server events in our adapter because this way we are sure to add the listener before the event is fired.

Reply:
I discussed this with the Prebid.js PMC and we'd prefer you put it into the analytics adapter. Is that possible?

Reply:
I understand and at the beginning we put it into the analytics. But in fact most of our customers do not want to compile analytics adapter and it is not a safe for us to move it.

Reply:
Any news @mkendall07 ?